### DISCUSSION TITLE
wrong number of arguments when trying to use batches

### DISCUSSION THREAD

**Discussion by mbajur** [bensheldon/good_job#1326] [General]
Hey! I'm pretty sure that's omething super stupid on my end but basically i am not able to use Batches feature in any way provided by README file. Using the simplest way:

```
class Feeds::FeedMigrationJob < ApplicationJob
  def perform
    GoodJob::Batch.enqueue do
      MyModel.find_each do |model|
        Feeds::MyModelFeedMigrationJob.perform_later(model.id)
      end
    end
  end
end
```

Is giving me this:

```
ArgumentError:
       wrong number of arguments (given 1, expected 0)
     # /Users/mbajur/.rvm/gems/ruby-3.2.3/gems/activerecord-7.0.8/lib/active_record/transactions.rb:208:in `transaction'
     # /Users/mbajur/.rvm/gems/ruby-3.2.3/gems/transaction_retry-1.0.3/lib/transaction_retry/active_record/base.rb:32:in `transaction_with_retry'
     # /Users/mbajur/.rvm/gems/ruby-3.2.3/gems/good_job-3.27.4/lib/good_job/adapter.rb:73:in `block in enqueue_all'
```

As long as i'm trying to provide some arguments to `Feeds::MyModelFeedMigrationJob.perform_later`. Once i remove the argument, it works (in specs at least) but this entire issue is only happening when trying to use batches. When i'm trying to do the same without batches, Feeds::MyModelFeedMigrationJob gets properly enqueued with arguments provided. Any thoughts what i might be doing wrong in here?

I'm on Rails 7.0 8 and good_job 3.27.4

Thanks!

**Comment by bensheldon** [comment]
I think the `transaction_retry` gem is not compatible with your version of Rails. 

This is how `transaction_retry` is calling Active Record's transaction (the method `transaction_without_retry` is aliased from `transaction` earlier in the file): 

`transaction_without_retry(*objects, &block)`

https://github.com/qertoip/transaction_retry/blob/0cd0491a5ef2a8e1864522bb2760536da039cb55/lib/transaction_retry/active_record/base.rb#L32

And here's how transaction is defined in Active Record 7.0: 

`def transaction(**options, &block)`

https://github.com/rails/rails/blob/1060cec88bfc5d45a6563d6966f6cb7533057b2b/activerecord/lib/active_record/transactions.rb#L208

The `transaction_retry` gem is calling `transaction` with args, the method only accepts kwargs. That gem also hasn't been updated in 6 years (that file in 11 years 😬).

That's probably only masking a different issue though, but I'd recommend you remove the gem and see what happens next.

**Reply by mbajur** [reply]
Ah, that's an excellent catch. I'm trying to add good_job to a very old rails app that recently got upgraded to rails 7. Thank you and i will try with your suggestion :)

**Reply by mbajur** [reply]
That was it! Removing the gem made everything working :) Thanks a ton for help and explanation and thank you for maintaining this gem
